common/short-hex-str/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use mirai_annotations::debug_checked_precondition;
+use serde::{Serialize, Serializer};
+use std::{fmt, str};
+use thiserror::Error;
+
+/// An efficient container for formatting a byte slice as a hex-formatted string,
+/// stored on the stack.
+///
+/// Using `ShortHexStr` instead of `hex::encode` is about 3-4x faster on a recent
+/// MBP 2019 (~48 ns/iter vs ~170 ns/iter) in an artifical micro benchmark.
+#[derive(Ord, PartialOrd, Eq, PartialEq, Hash, Clone, Copy)]
+pub struct ShortHexStr([u8; ShortHexStr::LENGTH]);
+
+#[derive(Error, Debug)]
+#[error("Input bytes are too short")]
+pub struct InputTooShortError;
+
+impl ShortHexStr {
+    pub const SOURCE_LENGTH: usize = 4;
+    pub const LENGTH: usize = 2 * ShortHexStr::SOURCE_LENGTH;
+
+    /// Format a new `ShortHexStr` from a byte slice.
+    ///
+    /// Returns `Err(InputTooShortError)` if the input byte slice length is less
+    /// than `SOURCE_LENGTH` bytes.
+    pub fn from_bytes(src_bytes: &[u8]) -> Result<ShortHexStr, InputTooShortError> {
+        if src_bytes.len() >= ShortHexStr::SOURCE_LENGTH {
+            Ok(ShortHexStr::from_bytes_unchecked(src_bytes))
+        } else {
+            Err(InputTooShortError)
+        }
+    }
+
+    /// Format a new `ShortHexStr` from a byte slice.
+    ///
+    /// Panics if the input byte slice length is less than `SOURCE_LENGTH` bytes.
+    pub fn from_bytes_unchecked(src_bytes: &[u8]) -> Self {
```

Comment:
This is called _unchecked which implies that this should be marked as unsafe.

Reply:
I think you're right. Changed to only `try_from_bytes(&[u8]) -> Result` and added `const_assert` to callers `HashValue::short_str` and `AccountAddress::short_str`.